app/models/user.rb

```diff
@@ -46,4 +47,7 @@ def authenticatable_salt
     self.read_attribute(Devise.saml_session_index_key.to_sym) if self.read_attribute(Devise.saml_session_index_key.to_sym).present?
   end
 
+  def get_role_by_grant(grant:)
+    self.grant_permission_role ||= self.system_admin? ? SYSTEM_ADMIN_GRANT_ROLE : GrantPermission.find_by(grant: grant, user: self).try(:role)
```

Comment:
if you set the grant permission role as a attribute on the user, and use it directly in the view, the role might be for a grant that's not the current one in context, and it might cause bugs. I suggest that you always call the get_role_by_grant(grant: foo) method and not use "self.grant_permission_role ||=" for now. If it becomes too slow, you can cache it by sth like this:

```
def get_role_by_grant(grant:)
    self.grant_permission_role[grant] ||= self.system_admin? ? SYSTEM_ADMIN_GRANT_ROLE : GrantPermission.find_by(grant: grant, user: self).try(:role)
```

Reply:
Ah, yes. This would cause an issue on the home and grants index pages. Thanks.

Reply:
Couldn't get the attr_accessor to work as an array. Instead, I tried re-setting the user's grant_permission_role per call. Doing that works, but doesn't seem right, either.